File: GLLara/GLLVertexAttribAccessor.m
//
//  GLLVertexAttribAccessor.m
//  GLLara
//
//  Created by Torsten Kammer on 06.04.21.
//  Copyright © 2021 Torsten Kammer. All rights reserved.
//

#import "GLLVertexAttribAccessor.h"

@implementation GLLVertexAttribAccessor

- (instancetype)initWithSemantic:(enum GLLVertexAttribSemantic)semantic layer:(NSInteger) layer format:(MTLVertexFormat)format dataBuffer:(NSData *__nullable)dataBuffer offset:(NSInteger)dataOffset stride:(NSInteger)stride;
{
    return [self initWithAttribute:[[GLLVertexAttrib alloc] initWithSemantic:semantic layer:layer format:format] dataBuffer:dataBuffer offset:dataOffset stride:stride];
}

- (instancetype)initWithAttribute:(GLLVertexAttrib *)attribute dataBuffer:(NSData *__nullable)dataBuffer offset:(NSInteger)dataOffset stride:(NSInteger)stride;
{
    if (!(self = [super init]))
        return nil;
    
    _attribute = attribute;
    _dataBuffer = dataBuffer;
    _dataOffset = dataOffset;
    
    if (stride == 0) {
        // As always in GL world, a stride of 0 means "calculate it yourself", not actually 0, which would make no sense.
        _stride = attribute.sizeInBytes;
    } else {
        _stride = stride;
    }
    
    return self;
}

- (NSInteger)offsetForElement:(NSInteger)index {
    return _dataOffset + index * _stride;
}

- (const void *)elementAt:(NSInteger)index {
    if (!_dataBuffer) return NULL;
    
    return _dataBuffer.bytes + [self offsetForElement:index];
}

- (NSData *)elementDataAt:(NSInteger)index {
    NSRange range = NSMakeRange([self offsetForElement:index], self.attribute.sizeInBytes);
    NSAssert(_dataBuffer != nil && NSMaxRange(range) <= _dataBuffer.length, @"Needs to be in range");
    return [_dataBuffer subdataWithRange:range];
}

@end

File: GLLara/GLLVertexAttribAccessor.swift
//
//  GLLVertexAttribAccessor.swift
//  GLLara
//
//  Created by Torsten Kammer on 04.07.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation

@objc class GLLVertexAttribAccessor: NSObject {
    
    @objc convenience init(semantic: GLLVertexAttribSemantic, layer: Int = 0, format: MTLVertexFormat, dataBuffer: Data?, offset: Int, stride: Int) {
        let attribute = GLLVertexAttrib(semantic: semantic, layer: layer, format: format)
        
        self.init(attribute: attribute, dataBuffer: dataBuffer, offset: offset, stride: stride)
    }
    
    init(attribute: GLLVertexAttrib, dataBuffer: Data?, offset: Int, stride: Int) {
        self.attribute = attribute
        self.dataBuffer = dataBuffer
        self.dataOffset = offset
        self.stride = stride
    }
    
    let attribute: GLLVertexAttrib
    let dataBuffer: Data?
    let dataOffset: Int
    let stride: Int
    
    func offset(element: Int) -> Int {
        return dataOffset + element * stride
    }
    
    func range(element: Int) -> Range<Int> {
        let start = offset(element: element)
        let length = attribute.sizeInBytes
        return start ..< (start + length)
    }
    
    func elementData(at element: Int) -> Data? {
        guard let dataBuffer = dataBuffer else {
            return nil
        }
        return try! dataBuffer.checkedSubdata(in: range(element: element))
    }
    
    func typedElementArray<T>(at element: Int, type: T.Type) -> [T] {
        guard let dataBuffer = dataBuffer else {
            assertionFailure()
            return []
        }
        
        return dataBuffer.withUnsafeBytes { start -> [T] in
            let slice = start[range(element: element)]
            let typed = UnsafeRawBufferPointer(rebasing: slice).bindMemory(to: T.self)
            return Array(typed)
        }
    }
    
    func simd2Element<T>(at element: Int, base: T.Type) -> SIMD2<T> where T: SIMDScalar {
        guard let dataBuffer = dataBuffer else {
            assertionFailure()
            return []
        }
        
        return dataBuffer.withUnsafeBytes { start -> SIMD2<T> in
            let slice = start[range(element: element)]
            let typed = UnsafeRawBufferPointer(rebasing: slice).bindMemory(to: T.self)
            return SIMD2<T>(x: typed[0], y: typed[1])
        }
    }
    
    func simd3Element<T>(at element: Int, base: T.Type) -> SIMD3<T> where T: SIMDScalar {
        guard let dataBuffer = dataBuffer else {
            assertionFailure()
            return []
        }
        
        return dataBuffer.withUnsafeBytes { start -> SIMD3<T> in
            let slice = start[range(element: element)]
            let typed = UnsafeRawBufferPointer(rebasing: slice).bindMemory(to: T.self)
            return SIMD3<T>(x: typed[0], y: typed[1], z: typed[2])
        }
    }
    
    func simd4Element<T>(at element: Int, base: T.Type) -> SIMD4<T> where T: SIMDScalar {
        guard let dataBuffer = dataBuffer else {
            assertionFailure()
            return []
        }
        
        return dataBuffer.withUnsafeBytes { start -> SIMD4<T> in
            let slice = start[range(element: element)]
            let typed = UnsafeRawBufferPointer(rebasing: slice).bindMemory(to: T.self)
            return SIMD4<T>(x: typed[0], y: typed[1], z: typed[2], w: typed[3])
        }
    }
    
    func withBytes(element: Int, action: (UnsafeRawBufferPointer) throws ->()) rethrows {
        guard let dataBuffer = dataBuffer else {
            assertionFailure()
            return
        }
        
        try dataBuffer.withUnsafeBytes { start in
            let slice = start[range(element: element)]
            try action(UnsafeRawBufferPointer(rebasing: slice))
        }
    }
}
